`typescript/no-floating-promises` behavior difference

I had Claude create a reproduction of the problem here, same config for ESLint and Oxlint, but ESLint passes while Oxlint raises a violation: https://github.com/connorshea/oxlint-sandbox/tree/claude/update-linting-typescript-U4ayT

You can see 2 errors in Oxlint, only 1 in ESLint:

```
oxlint-sandbox % pnpm oxlint

  × [typescript(no-floating-promises)](https://oxc.rs/docs/guide/usage/linter/rules/typescript/no-floating-promises.html): Promises must be awaited, add void operator to ignore.
   ╭─[repro/positive.ts:3:1]
 2 │
 3 │ getPromise();
   · ─────────────
   ╰────
  help: The promise must end with a call to .catch, or end with a call to .then with a rejection handler, or be explicitly marked as ignored with the `void` operator.

  × [typescript(no-floating-promises)](https://oxc.rs/docs/guide/usage/linter/rules/typescript/no-floating-promises.html): Promises must be awaited, add void operator to ignore.
   ╭─[repro/case.ts:9:1]
 8 │
 9 │ getTest();
   · ──────────
   ╰────
  help: The promise must end with a call to .catch, or end with a call to .then with a rejection handler, or be explicitly marked as ignored with the `void` operator.

Found 0 warnings and 2 errors.
Finished in 107ms on 4 files with 84 rules using 18 threads.

oxlint-sandbox % pnpm eslint

/Users/connorshea/code/oxlint-sandbox/repro/positive.ts
  3:1  error  Promises must be awaited, end with a call to .catch, end with a call to .then with a rejection handler or be explicitly marked as ignored with the `void` operator  @typescript-eslint/no-floating-promises

✖ 1 problem (1 error, 0 warnings)
```

Code in question (https://github.com/connorshea/oxlint-sandbox/blob/claude/update-linting-typescript-U4ayT/repro/case.ts):
```ts
interface MyPromise<T> extends Promise<T> {
  __brand: string;
}

type Test = MyPromise<number> & { foo: 'bar' };

declare function getTest(): Test;

getTest();
```

This is the config that triggers the bug, I think something is wrong with `allowForKnownSafePromises`:
```json
{
  "$schema": "./node_modules/oxlint/configuration_schema.json",
  "options": {
    "typeAware": true
  },
  "plugins": ["typescript"],
  "rules": {
    "typescript/no-floating-promises": [
      "error",
      {
        "allowForKnownSafePromises": [
          { "from": "file", "name": "MyPromise" }
        ]
      }
    ]
  }
}
```

--- a/internal/rules/no_floating_promises/no_floating_promises_test.go
+++ b/internal/rules/no_floating_promises/no_floating_promises_test.go
@@ -786,6 +786,17 @@ declare function createMyThenable(): MyThenable;
 
 createMyThenable();
     `},
+		{
+			Code: `
+interface SafePromise<T> extends Promise<T> {
+  __brand: string;
+}
+type BrandedSafePromise = SafePromise<number> & { foo: 'bar' };
+declare function getPromise(): BrandedSafePromise;
+getPromise();
+      `,
+			Options: rule_tester.OptionsFromJSON[NoFloatingPromisesOptions](`{"allowForKnownSafePromises": [{"from": "file", "name": "SafePromise"}]}`),
+		},
 	}, []rule_tester.InvalidTestCase{
 		{
 			Code: `

--- a/internal/utils/type_matches_specifier.go
+++ b/internal/utils/type_matches_specifier.go
@@ -474,13 +474,24 @@ func TypeMatchesSomeSpecifier(
 	specifiers []TypeOrValueSpecifier,
 	program *compiler.Program,
 ) bool {
+	matches := func(t *checker.Type) bool {
+		if IsIntrinsicErrorType(t) {
+			return false
+		}
+		return Some(specifiers, func(s TypeOrValueSpecifier) bool {
+			return typeMatchesSpecifier(t, s, program)
+		})
+	}
+
+	if matches(t) {
+		return true
+	}
+
 	for _, typePart := range IntersectionTypeParts(t) {
-		if IsIntrinsicErrorType(typePart) {
+		if typePart == t {
 			continue
 		}
-		if Some(specifiers, func(s TypeOrValueSpecifier) bool {
-			return typeMatchesSpecifier(t, s, program)
-		}) {
+		if matches(typePart) {
 			return true
 		}
 	}

